### ISSUE TITLE
AD Security Group sync does not work when group has large number of members

### ISSUE THREAD

**Issue by afwagner** [adobe-apiplatform/user-sync.py#423]
user-sync.pex 2.4rc1
Windows Server 2016
Python 3.6.5

AD Security Group sync does not work when group has large number of members (over 13,000).
No errors are reported regarding the sync, the Adobe group is not filled with any members.

Syncing AD Security Groups with smaller number of members (under 2,000) works.

Note: In the .NET Framework world, I was able to resolve this problem by moving Security Group Management to the newer System.​Directory​Services.​Account​Management calls.  I realize this application is written in Python.  Is there a newer set of AD Libraries that can be used?

**Comment by adorton-adobe** [comment]
Have you modified any of the queries in the LDAP config?  Will you please send me your LDAP config (with credentials stripped) and your user sync config?  You can email them to adorton@adobe.com if you don't want to post them here.

**Comment by adorton-adobe** [comment]
Also - I don't recommend using 2.4rc1 unless you need to use the new group sync features.  I strongly recommend using 2.3 until 2.4 stabilizes.

**Comment by afwagner** [comment]
Andrew,

Here are the two config files.

I have modified the all_users_filter in connector-ldap.yml.  It's producing
the desired results regarding users I want to sync up with Adobe.

In the user-sync-config.yml file "groups" settings:
SDWAllStaff and SDWAllTeachers are syncing, The SDWStudents group is not.
The first two groups are under 2000 members each, where the SDWStudents
group is over 13000 members.

Alan

**Comment by adorton-adobe** [comment]
@afwagner Were you able to get this resolved?  If not, will you please email me your config files directly?  I won't get them if you attach them to an email reply to this Github issue.

Did you also try running v2.3?

**Comment by afwagner** [comment]
Andrew,

 

Thanks for getting back. No, the problem is not resolved.

 

Of the AD Groups in configuration file user-sync-config.yml, only “SDWStudents” does not fully sync.  It’s very large, with over 13,000 entries.  As referenced in the ticket, I’ve had this problem with the legacy LDAP API’s in other applications I write using .NET.  When I switched over to the newer System.​Directory​Services.​Account​Management calls the problem went away.

 

   - directory_group: "SDWAllStaff"

      adobe_groups:

        - "SDWAllStaff"

    - directory_group: "SDWAllTeachers"

      adobe_groups:

        - "SDWAllTeachers"

    - directory_group: "SDWStudents"

      adobe_groups:

        - "SDWStudents"

    - directory_group: "Domain Admins"

      adobe_groups:

        - "SDWDomainAdmins"

 

Thanks again,

Alan Wagner

Systems Programmer / Data Base Administrator

School District Of Waukesha

afwagner@waukesha.k12.wi.us

(262) 970-1056

**Comment by adorton-adobe** [comment]
This might take some time to investigate.  I'll keep you posted with any updates or if I need additional information.

Are you still running the release candidate?  If so, will you please try running it with 2.3 stable and let me know if you experience the same issue?

**Comment by Luci2015** [comment]
Just FYI, I had recently a situation when there were partial results obtained from AD, from big groups (over 10k accounts inside).
The fix this time was using an account that had extended rights.
If you try to set up the tool with your username (or at least the one you get +10k results for in AD) and also change the "search_page_size" in connector-ldap.yml file to 200, do you see a change?

**Comment by afwagner** [comment]
Lucian,

Thanks for your feedback.  Tried both items you mentioned.  Same issue as originally stated.

Alan

**Comment by SCRICrtoole** [comment]
user-sync 2.4 on RHEL

I am having an almost identical issue.  Our staff groups (1,200+ members max) are syncing perfectly, while student groups (2549+ members) are not returned.

I checked the logs on the domain controller, no errors.  Also no errors in the user-sync debug output.

2019-06-19 11:04:56 1352 DEBUG ldap - Count of users in group "student group name here": 0

**Comment by Luci2015** [comment]
The tool works pretty well in most of the deployments I've faced. In the problematic ones (that returned 0 results for specific groups), they have been cleared by local LDAP setup or usual misconfiguration of the connector-ldap file.
Here is a quick list of error sources:
(config files cause)
- base_dn is too restrictive-> users and or groups can't be discovered in provided path
- contacted host is a domain controller not up to date (force replication?)
- default queries in the connector-ldap file have been modified
- pagination is set to 0 instead of some integer value (ex: 200 or 1000)
(local ldap cause)
- the service account that the tool uses to extract from LDAP does not have enough permissions
- the groups to be extracted are part of different OU's and one of them might be missing some default permissions in security tab
- there is a limit imposed in the LDAP for the amount of results to be obtained
- using global catalog port to extract objects from sub-domains the target groups are not set to Universal and the base_dn in the config file takes a value different than ""
- memberOf attribute is missing
Possibly many others from local LDAP setup

Usually if one groups works and the other does not, you already have a starting point.
IMO a comparison between permissions/settings/limits on both (or their parent OU's) could lead to a resolve.
If there is a local limit set, you can find it by trial and error by adding /removing until you get that number and search inside the settings until you find something that matches that number..

**Comment by adorton-adobe** [comment]
I agree with @Luci2015 - the size of the group shouldn't be an issue.  Definitely not for 2.4k or even 13k (as mentioned in the issue).  I've seen setups that can sync tens of thousands of users in the single group with no issues.

I would also add that it would be worth upgrading to v2.5.0 - it uses a different ldap client library.

**Comment by SCRICrtoole** [comment]
Thanks for the follow-up!

When searching using the same user account and baseDN I get results for both working and non working groups.  

I double cheked the points mentioned with the following results.

* All groups are configured identically in the user-sync-config.yml.  8 working 1 not working
* Domain controllers are all fully up to date
* Connector-LDAP values look fine:
             * base_dn: "dc=domain,dc=com"
             * all_users_filter: "(&(objectClass=user)(objectCategory=person))"
             * group_filter_format: "(&(objectCategory=group)(cn={group}))"
             * group_member_attribute_name: "member"
* search_page_size: 1000 (I have tried setting this lower)
* May be a local limit imposed by LDAP: this is possible but would be at the default in our case currently
* OU permissions are correct, no inheritcance blocked or anything similar.
* using LDAPS port, not global catalog
* memberOf attribute is populated

Any other ideas of what I should check, or should I follow down the road to see what the default limit is in AD?  Its been a long time, but i seem to recall that being somewhere around 5000.

We have multiple instances of Adobe-sync running.  Different domains, same issue.

Thanks!

**Comment by adorton-adobe** [comment]
@SCRICrtoole Can you try v2.5.0 and see if you still have the issue?

**Comment by SCRICrtoole** [comment]
@adorton-adobe I will try 2.5.0 tomorrow and update with the results.

**Comment by SCRICrtoole** [comment]
I have updated both instances to 2.5.0 and ran a sync with --test-mode.  All users including students were found and flagged to be provisioned.  

Tomorrow I will finish reviewing the logs and run without the --test-mode switch and let you know what happens.

thanks for the help

**Comment by SCRICrtoole** [comment]
@adorton-adobe @Luci2015 Thanks for your help guys!  The sync is now working with the groups in question after updating to v2.5.0.

**Comment by adorton-adobe** [comment]
Excellent.  Thank you for letting us know.  I'm closing this since we appear to have a clear solution to the issue.  @afwagner please try using v2.5.0 and let me know if that does not resolve the issue.

**Comment by Luci2015** [comment]
post closure, in case it helps somebody else that get partial results from AD

TLDR:  PrimaryGroup names used in group mapping can reduce the number of users to be synced

EX:
(run in poweshell: `Get-ADUser user1 -Properties *` then pick up the PrimaryGroup value)
--> say user1 has the PrimaryGroup=`CN=Domain Users,CN=Users,DC=domain,DC=com`
MemberOf tab shows the user is member of `Domain Users` and of `Students` group.
--> say user2 has PrimaryGroup=`CN=Students,OU=Security,DC=domain,DC=com`
MemberOf tab shows the user is member of `Domain Users` and of `Students` group

case1:  assume we have this mapping where the PrimaryGroup `Students` is being targeted for members to be synced over to Admin Console
```
  groups:
    - directory_group: "Students"
      adobe_groups:
        - "active"
```
Normal run of the tool with `--users mapped` will not pick up user2, nor any other user that has the PrimaryGroup set to `CN=Students,OU=Security,DC=domain,DC=com`
user1 is being picked, though, as it is a memberOf `CN=Students`

case2:  assume we have this mapping
```
  groups:
    - directory_group: "Domain Users"
      adobe_groups:
        - "active"
```
Normal run of the tool with `--users mapped` will not pick up user1, nor any other user that has the PrimaryGroup set to `CN=Domain Users,CN=Users,DC=domain,DC=com` but will pick up all the users that have memberOf `CN=Domain Users` AND a different PrimaryGroup than `Domain Users`

Tip for a workaround in case1:
find the primaryGroupID with the same `Get-ADUser user1 -Properties *` command line then inside connector-ldap.yml file try the found value (I'll use `81841` as the primaryGroupID  of `Students` group):

`group_member_filter_format: "(|(primaryGroupID=81841)(memberOf={group_dn}))"`

This extracts:
* all the users that have `CN=Students,OU=Security,DC=domain,DC=com` as their PrimaryGroup
* all the users that are members of `CN=Students` AND have different PrimaryGroup.

Of course the `CN=Students,OU=Security,DC=domain,DC=com` should be cleaned up from any unnecessary accounts before syncing over to Admin Console.

**The PrimaryGroup name can be extended to ANY builtin AD group (ex: `Domain Admins`, `Domain Guests` etc...) =  same effect will be noticed for their members.**
